lib/shopify_api/rest.ex

```diff
@@ -11,25 +11,81 @@ defmodule ShopifyAPI.REST do
   alias ShopifyAPI.JSONSerializer
   alias ShopifyAPI.REST.Request
 
-  @doc false
-  def get(%AuthToken{} = auth, path, params \\ []) do
-    Request.perform(auth, :get, path, "", params)
+  @doc """
+  Underlying utility retrieval function. The options passed affect both the
+  return value and, ultimately, the number of requests made to Shopify.
+
+  ## Options
+
+    `:pagination` - Can be `:none`, `:stream`, or `:auto`. Defaults to :auto
+    `:auto` will block until all the pages have been retrieved and concatenated together.
+    `:none` will only return the first page. You won't have access to the headers to manually
+      paginate.
+    `:stream` will return a `Stream`, prepopulated with the first page.
+  """
+  @spec get(AuthToken.t(), path :: String.t(), keyword(), keyword()) ::
+          {:ok, %{required(String.t()) => [map()]}} | Enumerable.t()
+  def get(%AuthToken{} = auth, path, params \\ [], options \\ []) do
+    case pagination(options) do
+      :none ->
+        with {:ok, response} <- Request.perform(auth, :get, path, "", params) do
+          {:ok, fetch_body(response)}
+        end
+
+      :stream ->
+        Request.stream(auth, path, params)
+
+      :auto ->
+        auth
+        |> Request.stream(path, params)
+        |> collect_results()
+    end
+  end
+
+  @spec collect_results(Enumerable.t()) ::
+          {:ok, list()} | {:error, HTTPoison.Response.t() | any()}
+  defp collect_results(stream) do
+    stream
+    |> Enum.reduce_while({:ok, []}, fn
+      %{} = result, {:ok, acc} -> {:cont, {:ok, [result | acc]}}
+      error, {:ok, _acc} -> {:halt, error}
```

Comment:
🛠 This currently assumes each entry in a paginated response will be a map. This _feels_ pretty truthy to me, but maybe a bit safer if we flip the clauses and explicitly match on errors?

```elixir
Enum.reduce_while(stream, {:ok, []}, fn
  {:error, _} = error, {:ok, _acc} -> {:halt, error}
  result, {:ok, acc} -> {:cont, {:ok, [result | acc]}}
end)
```